File: lib/delivery-radius.ts
/**
 * Delivery zone matching — polygon zones preferred; circle tiers as fallback.
 * Pure helpers — safe for client and server.
 */

import {
  formatRadiusEta,
  type DeliveryRadiusTier,
  type DeliveryZone,
  type DeliveryZonePoint,
} from '@/lib/app-settings-schema'
import { calculateDistance } from '@/lib/utils'

function roundMoney(value: number): number {
  if (!Number.isFinite(value)) return 0
  return Math.round(value * 100) / 100
}

export type DeliveryZoneMatch =
  | {
      inRange: true
      tier: DeliveryRadiusTier | null
      zone: DeliveryZone | null
      distanceKm: number
      feeAmount: number
      estimate: string
      etaMinMinutes: number
      etaMaxMinutes: number
      label: string
    }
  | {
      inRange: false
      distanceKm: number
      maxRadiusKm: number
      feeAmount: null
      estimate: null
    }

/** Haversine distance in kilometres. */
export function distanceKmBetween(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  return calculateDistance(lat1, lon1, lat2, lon2) / 1000
}

/** Ray-casting point-in-polygon (lat/lng treated as planar for local zones). */
export function pointInPolygon(
  point: DeliveryZonePoint,
  polygon: DeliveryZonePoint[]
): boolean {
  if (polygon.length < 3) return false
  let inside = false
  for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    const xi = polygon[i].lng
    const yi = polygon[i].lat
    const xj = polygon[j].lng
    const yj = polygon[j].lat
    const intersect =
      yi > point.lat !== yj > point.lat &&
      point.lng < ((xj - xi) * (point.lat - yi)) / (yj - yi + 1e-15) + xi
    if (intersect) inside = !inside
  }
  return inside
}

/** Approximate polygon area in deg² (relative only — for smallest-zone pick). */
export function polygonAreaDeg2(polygon: DeliveryZonePoint[]): number {
  if (polygon.length < 3) return Number.POSITIVE_INFINITY
  let sum = 0
  for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
    sum += polygon[j].lng * polygon[i].lat - polygon[i].lng * polygon[j].lat
  }
  return Math.abs(sum) / 2
}

export function matchDeliveryPolygonZone(
  zones: DeliveryZone[],
  lat: number,
  lng: number
): DeliveryZone | null {
  if (!zones.length || !Number.isFinite(lat) || !Number.isFinite(lng)) return null
  const point = { lat, lng }
  let best: DeliveryZone | null = null
  let bestArea = Number.POSITIVE_INFINITY
  for (const zone of zones) {
    if (!pointInPolygon(point, zone.polygon)) continue
    const area = polygonAreaDeg2(zone.polygon)
    if (area < bestArea) {
      bestArea = area
      best = zone
    }
  }
  return best
}

/** First tier whose maxRadiusKm covers [distanceKm], or null if outside all zones. */
export function matchDeliveryRadiusTier(
  tiers: DeliveryRadiusTier[],
  distanceKm: number
): DeliveryRadiusTier | null {
  if (!tiers.length || !Number.isFinite(distanceKm) || distanceKm < 0) return null
  for (const tier of tiers) {
    if (distanceKm <= tier.maxRadiusKm + 1e-9) return tier
  }
  return null
}

export function maxDeliveryRadiusKm(tiers: DeliveryRadiusTier[]): number {
  if (!tiers.length) return 0
  return Math.max(...tiers.map((tier) => tier.maxRadiusKm))
}

/** Circle-tier matching (legacy). */
export function resolveDeliveryZone(params: {
  tiers: DeliveryRadiusTier[]
  storeLat: number
  storeLng: number
  addressLat: number
  addressLng: number
}): DeliveryZoneMatch | null {
  return resolveDeliveryCoverage({
    zones: [],
    tiers: params.tiers,
    storeLat: params.storeLat,
    storeLng: params.storeLng,
    addressLat: params.addressLat,
    addressLng: params.addressLng,
  })
}

/**
 * Prefer polygon zones when present; otherwise fall back to circle tiers.
 * Returns null when neither is configured.
 */
export function resolveDeliveryCoverage(params: {
  zones?: DeliveryZone[] | null
  tiers?: DeliveryRadiusTier[] | null
  storeLat: number
  storeLng: number
  addressLat: number
  addressLng: number
}): DeliveryZoneMatch | null {
  const zones = params.zones ?? []
  const tiers = params.tiers ?? []
  const { storeLat, storeLng, addressLat, addressLng } = params

  if (
    !Number.isFinite(storeLat) ||
    !Number.isFinite(storeLng) ||
    !Number.isFinite(addressLat) ||
    !Number.isFinite(addressLng)
  ) {
    return null
  }

  const distanceKm = distanceKmBetween(storeLat, storeLng, addressLat, addressLng)

  if (zones.length > 0) {
    const zone = matchDeliveryPolygonZone(zones, addressLat, addressLng)
    if (!zone) {
      return {
        inRange: false,
        distanceKm,
        maxRadiusKm: 0,
        feeAmount: null,
        estimate: null,
      }
    }
    return {
      inRange: true,
      tier: null,
      zone,
      distanceKm,
      feeAmount: roundMoney(zone.feeAmount),
      estimate: formatRadiusEta(zone.etaMinMinutes, zone.etaMaxMinutes),
      etaMinMinutes: zone.etaMinMinutes,
      etaMaxMinutes: zone.etaMaxMinutes,
      label: zone.label,
    }
  }

  if (!tiers.length) return null

  const tier = matchDeliveryRadiusTier(tiers, distanceKm)
  const maxRadiusKm = maxDeliveryRadiusKm(tiers)

  if (!tier) {
    return {
      inRange: false,
      distanceKm,
      maxRadiusKm,
      feeAmount: null,
      estimate: null,
    }
  }

  return {
    inRange: true,
    tier,
    zone: null,
    distanceKm,
    feeAmount: roundMoney(tier.feeAmount),
    estimate: formatRadiusEta(tier.etaMinMinutes, tier.etaMaxMinutes),
    etaMinMinutes: tier.etaMinMinutes,
    etaMaxMinutes: tier.etaMaxMinutes,
    label: tier.label,
  }
}

File: upaharo_mobile/lib/core/utils/delivery_radius.dart
import 'dart:math' as math;

/// One admin-configured delivery zone from the store pin.
class DeliveryRadiusTier {
  const DeliveryRadiusTier({
    required this.id,
    required this.maxRadiusKm,
    required this.feeAmount,
    required this.etaMinMinutes,
    required this.etaMaxMinutes,
    this.label = '',
  });

  final String id;
  final double maxRadiusKm;
  final double feeAmount;
  final int etaMinMinutes;
  final int etaMaxMinutes;
  final String label;

  String get estimateLabel {
    final lo = etaMinMinutes < 1 ? 1 : etaMinMinutes;
    final hi = etaMaxMinutes < lo ? lo : etaMaxMinutes;
    if (lo == hi) return 'Estimated delivery: $lo minutes';
    return 'Estimated delivery: $lo-$hi minutes';
  }

  factory DeliveryRadiusTier.fromJson(Map<String, dynamic> json) {
    final minEta = (json['etaMinMinutes'] as num?)?.round() ?? 20;
    var maxEta = (json['etaMaxMinutes'] as num?)?.round() ?? 30;
    if (maxEta < minEta) maxEta = minEta;
    return DeliveryRadiusTier(
      id: json['id'] as String? ?? '',
      maxRadiusKm: (json['maxRadiusKm'] as num?)?.toDouble() ?? 0,
      feeAmount: (json['feeAmount'] as num?)?.toDouble() ?? 0,
      etaMinMinutes: minEta,
      etaMaxMinutes: maxEta,
      label: json['label'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'maxRadiusKm': maxRadiusKm,
        'feeAmount': feeAmount,
        'etaMinMinutes': etaMinMinutes,
        'etaMaxMinutes': etaMaxMinutes,
        'label': label,
      };
}

/// Result of matching an address against configured zones.
class DeliveryZoneMatch {
  const DeliveryZoneMatch.inRange({
    required this.tier,
    required this.distanceKm,
  }) : inRange = true,
       maxRadiusKm = null;

  const DeliveryZoneMatch.outOfRange({
    required this.distanceKm,
    required this.maxRadiusKm,
  }) : inRange = false,
       tier = null;

  final bool inRange;
  final DeliveryRadiusTier? tier;
  final double distanceKm;
  final double? maxRadiusKm;

  double get feeAmount => tier?.feeAmount ?? 0;
  String? get estimate => tier?.estimateLabel;
}

double distanceKmBetween(
  double lat1,
  double lon1,
  double lat2,
  double lon2,
) {
  const earthKm = 6371.0;
  final dLat = _rad(lat2 - lat1);
  final dLon = _rad(lon2 - lon1);
  final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
      math.cos(_rad(lat1)) *
          math.cos(_rad(lat2)) *
          math.sin(dLon / 2) *
          math.sin(dLon / 2);
  final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
  return earthKm * c;
}

double _rad(double deg) => deg * math.pi / 180;

DeliveryRadiusTier? matchDeliveryRadiusTier(
  List<DeliveryRadiusTier> tiers,
  double distanceKm,
) {
  if (tiers.isEmpty || distanceKm.isNaN || distanceKm < 0) return null;
  final sorted = [...tiers]..sort((a, b) => a.maxRadiusKm.compareTo(b.maxRadiusKm));
  for (final tier in sorted) {
    if (distanceKm <= tier.maxRadiusKm + 1e-9) return tier;
  }
  return null;
}

DeliveryZoneMatch? resolveDeliveryZone({
  required List<DeliveryRadiusTier> tiers,
  required double storeLat,
  required double storeLng,
  required double addressLat,
  required double addressLng,
}) {
  if (tiers.isEmpty) return null;
  if (addressLat == 0 && addressLng == 0) return null;

  final distanceKm =
      distanceKmBetween(storeLat, storeLng, addressLat, addressLng);
  final tier = matchDeliveryRadiusTier(tiers, distanceKm);
  if (tier == null) {
    final maxKm = tiers
        .map((t) => t.maxRadiusKm)
        .fold<double>(0, (a, b) => a > b ? a : b);
    return DeliveryZoneMatch.outOfRange(
      distanceKm: distanceKm,
      maxRadiusKm: maxKm,
    );
  }
  return DeliveryZoneMatch.inRange(tier: tier, distanceKm: distanceKm);
}

List<DeliveryRadiusTier> parseDeliveryRadiusTiers(dynamic raw) {
  if (raw is! List) return const [];
  final out = <DeliveryRadiusTier>[];
  final seen = <int>{};
  for (final item in raw) {
    if (item is! Map) continue;
    final tier =
        DeliveryRadiusTier.fromJson(Map<String, dynamic>.from(item));
    if (tier.maxRadiusKm < 0.1) continue;
    final key = (tier.maxRadiusKm * 100).round();
    if (seen.contains(key)) continue;
    seen.add(key);
    out.add(tier);
    if (out.length >= 12) break;
  }
  out.sort((a, b) => a.maxRadiusKm.compareTo(b.maxRadiusKm));
  return out;
}
